File: tech-stack-implementations/backend/container-python-fastapi/src/app/api/health.py
"""
App: container-python-fastapi
Package: src.app.api
File: health.py
Version: 0.1.0
Turns: 6
Author: AI Coding Agent (Claude Opus 4.5)
Date: 2026-03-28T20:15:00Z
Exports: router
Description: Health check endpoints for liveness and readiness probes
Log:
6, 0.1.0, 2026/03/28, 08:15 PM, Claude Opus 4.5
"""

from datetime import datetime

from fastapi import APIRouter
from pydantic import BaseModel

from src.app.core.config import get_settings
from src.app.models.response import ApiResponse

router = APIRouter(prefix="/health", tags=["Health"])


class HealthStatus(BaseModel):
    status: str
    service: str
    version: str
    environment: str
    timestamp: datetime


class ReadinessStatus(BaseModel):
    status: str
    checks: dict[str, str]


@router.get("/live", response_model=ApiResponse[HealthStatus])
async def liveness() -> ApiResponse[HealthStatus]:
    settings = get_settings()
    return ApiResponse(
        data=HealthStatus(
            status="healthy",
            service=settings.app_name,
            version="0.1.0",
            environment=settings.app_env,
            timestamp=datetime.utcnow(),
        )
    )


@router.get("/ready", response_model=ApiResponse[ReadinessStatus])
async def readiness() -> ApiResponse[ReadinessStatus]:
    return ApiResponse(
        data=ReadinessStatus(
            status="ready",
            checks={
                "self": "ok",
            },
        )
    )

File: tech-stack-implementations/backend/serverless-typescript-lambda/src/handlers/health.ts
/**
 * App: serverless-typescript-lambda
 * Package: src/handlers
 * File: health.ts
 * Version: 0.1.0
 * Task: 001
 * Turns: 001
 * Author: AI Coding Agent (claude-opus-4-5-20251101)
 * Date: 2026-03-30T00:00:00Z
 * Exports: handler
 * Description: Health check Lambda handler for liveness and readiness probes
 * Log:
 * 001, 001, 0.1.0, 2026/03/30, 12:00 AM, AI Coding Agent (claude-opus-4-5-20251101)
 */

import type { APIGatewayProxyEventV2, APIGatewayProxyStructuredResultV2 } from 'aws-lambda';
import type { HealthStatus } from '../types/api.js';
import { logger, metrics, tracer, MetricUnit } from '../middleware/observability.js';
import { createHandlerContext } from '../middleware/correlation.js';
import * as response from '../utils/response.js';
import { handleError } from '../utils/errors.js';

const APP_VERSION = process.env.APP_VERSION ?? '0.1.0';

export const handler = async (
  event: APIGatewayProxyEventV2
): Promise<APIGatewayProxyStructuredResultV2> => {
  const segment = tracer.getSegment();
  const subsegment = segment?.addNewSubsegment('health-handler');

  const ctx = createHandlerContext(event);

  try {
    logger.info('Health check request', { path: ctx.path });

    const healthStatus: HealthStatus = {
      status: 'healthy',
      version: APP_VERSION,
      timestamp: new Date().toISOString(),
    };

    if (ctx.path === '/health/ready') {
      healthStatus.checks = {
        database: { status: 'pass' },
      };
    }

    metrics.addMetric('HealthChecks', MetricUnit.Count, 1);
    metrics.addDimension('path', ctx.path);
    metrics.publishStoredMetrics();

    return response.ok(healthStatus, { correlationId: ctx.correlationId });
  } catch (error) {
    return handleError(error, ctx.correlationId);
  } finally {
    subsegment?.close();
  }
};
